Add color pickers to color equalizer

jenshannoschwalm opened the pull request:
Add a picker togglebutton to the hue_shift slider. The quality of box picker statistics for hue certainly needs further care in `dt_color_picker_helper()` but is available here.

All live-pickers plus the in-module picker are visualized.

Fixes #16265

elstoc commented:
Looks nice. Would it be possible to keep the picker active while changing the module parameters? At the moment it needs to be re-selected each time a module parameter is changed

jenshannoschwalm commented:
> Would it be possible to keep the picker active while changing the module parameters?

Honestly i don't know right now whether i did something wrong or it's due to the picker implementation.

 

> Also the area picker seems somewhat odd.

I think you mean the "range" here. Yes, that's really odd. I mentioned this by "quality of box picker statistics for hue", that's how statistics for hue are done atm, we would have to do some circular code for hue. There are some FIXMEs in the code, as i hope i can fix those i offered the area picker here.

elstoc commented:
I think the general principle with the picker is that when it's used just to show information (as in the tone curve module) it should remain active, but when it's used to control the value of a slider then it should be de-activated when any module control is amended, otherwise odd things will happen.

With the area/range picker, if you select an area that covers multiple colours, then unless those colours are adjacent to one another (in hue) then the only practical way to display would be multiple disconnected ranges (which doesn't really make much sense), or to just take an average (which is maybe better, but still not particularly useful). I'd probably opt for disabling the area/range picker in this module.

jenshannoschwalm commented:
@elstoc @TurboGit 
1. The picker getting inactivated as reported by chris has been fixed, two subtle errors. squashed and force-pushed.
2. About the area picker. It's working well already for areas that a) don't include completely achromatic parts and b) don't include reds as we don't do circular correction. So i would opt to keep areas. Will certainly try to fix deeper in the colorpicker code...

MStraeten commented:
found some weird behaviour when used in combination with formerly defined presets: color line indicates blue in a proper reddish skin - need to investigate a bit to get a reproducible simple scenario... 
here a demo with a colorchecker ...
https://github.com/darktable-org/darktable/assets/39386816/619e8c96-36b2-43a6-99f1-7c2fccd295b3

jenshannoschwalm commented:
> found some weird behaviour when used in combination with formerly defined presets: color line indicates blue in a proper reddish skin - need to investigate a bit to get a reproducible simple scenario...

Yes, that's the issue discussed. Hue mean (and min/max) is not properly calculated (see https://stackoverflow.com/questions/491738/how-do-you-calculate-the-average-of-a-set-of-circular-data ) in current picker code.

jenshannoschwalm commented:
> here a demo with a colorchecker ...

hmmm ...

jenshannoschwalm commented:
@MStraeten confirmed. Something strange going on here, don't know yet.

jenshannoschwalm commented:
1. Fixed the issue reported by @MStraeten 
2. Increased colorfulness of gui elements for those with imperfect eyes
3. improved picker markers
4. Added some picker pixelpipe debugging info i found useful for tracking the reported issue

Seems to be fine now @TurboGit 

Just a reminder, there seem to be a number of subtle issues with the colorpicker that need attention
1. why is the picker colorspace changed in the background while using a preset or reset default? 
2. we loose a lot of performance as we disable the pipecache whenever we move a picker. Likely we can just drop some cachelines
3. the picker denoising could certainly be improved for performance as a point picker certainly does not require the full area to be blurred.
4. the live picker always present pixelpipe output instead of module input

elstoc commented:
Looking good now, thanks. Looking forward to making some nice B&W images once this lands in stable - gives so much more control than I can get with color calibration.

jenshannoschwalm commented:
> making some nice B&W images once this lands in stable

What module are you using to do the B&W conversion? Good old "monochrome"?

jenshannoschwalm commented:
> Could we have `right-clic` as a possible way to get a picker area as in `rgb curve`?

ATM i don't think so. At least not if we want to keep the picker non-obtrusive as here inside the line as that mode uses a toggle-button. That doesn't report a `GdkEventButton` to `_color_picker_callback_button_press()` thus this doesn't work.

todd-prior commented:
I just compiled DT on Windows last night...I don't see any picker or toggle... I will look again  Can someone post a screen shot of what it looks like... the only one I see is for white level.....

elstoc commented:
It's next to the node placement slider, immediately below the graph

todd-prior commented:
![image](https://github.com/darktable-org/darktable/assets/44409419/0bbb2aa0-66b6-4ff0-b239-48ab964df8ba)

In the hue tab??

elstoc commented:
Nope you must have a version before the picker was merged. Try pulling and rebuilding again.

todd-prior commented:
> Nope you must have a version before the picker was merged. Try pulling and rebuilding again.

No it was current even as of today...but I had not deleted my build directory last night or today when I tried ... so once I did that and recompiled it appeared... sorry for the noise.....
